### PR TITLE
feat(headless): added case classification-related actions

### PR DESCRIPTION
**Pull request by nathanlb**
Added CaseInput actions:
- setCaseInput

Added CaseFieldsActions:
- setCaseField
- enableCaseClassifications
- disableCaseClassifications
- fetchCaseClassifications

### CODE REVIEW

**Comment by nathanlb** [comment]
Still in [DRAFT] due to missing analytics actions which are in progress.

File: packages/headless/src/api/service/case-assist/case-assist-params.test.ts
Comment (lbergeron) [inline packages/headless/src/api/service/case-assist/case-assist-params.test.ts:68]: When the API client was written, it was intentional to set the value directly on the property. The purpose was to make it easier to use by removing the unnecessary complexity of the `value` level.

Please check if `CaseAssistApiClient` was exposed to end users before your changes. If yes, then the change would be considered breaking.

**Review by lbergeron — COMMENTED** [review]
Looks good overall. But please double check `CaseAssistApiClient` to make sure we're not introducing breaking changes.

File: packages/headless/src/api/service/case-assist/case-assist-params.test.ts
Comment (nathanlb) [reply packages/headless/src/api/service/case-assist/case-assist-params.test.ts:68]: It may be slightly easier to use in the unlikely scenario that someone is using the client without any controllers right now.

But when using the client with values from the state, which is 99% of the case, it forces us to parse the state into this previous format to pass as arguments, then the arguments get parsed again immediately into exactly the same shape as before in order to be sent to the API.

Considering the state is holding this data and the API request parameters have the same shape. I think this is more suitable.

I don't see the CaseAssistAPIClient being explicitly exposed but if you could be my second opinion on this it would be greatly appreciated. Thanks!

File: packages/headless/src/features/case-fields/case-fields-slice.ts
Comment (olamothe) [inline packages/headless/src/features/case-fields/case-fields-slice.ts:27]: What are the scenarios where you would want to enable or disable case classification, and what impact should it have on the internal logic (if any) ?

For example, if case assis is disabled, and someone dispatch `fetchCaseClassifications` action, what should happen ?

Right now I'm not sure I understand what the `enabled` flag does exactly.

**Review by olamothe — APPROVED** [review]
:gg:

File: packages/headless/src/features/case-fields/case-fields-slice.ts
Comment (nathanlb) [reply packages/headless/src/features/case-fields/case-fields-slice.ts:27]: So this "enabled" flag activates the **automatic** retrieval of case classifications when changes are made to the the `caseInput` values. You can still disable it and fetch by dispatching the fetch action but the intended way of using it is to register caseInputs (ex: subject and description) and instantiate a case classification controller that enables this automatic retrieval feature and will react to changes in the inputs.

File: packages/headless/src/features/index.ts
Comment (samisayegh) [inline packages/headless/src/features/index.ts:32]: I believe these exports will be exported in the search use-case.

Suggestion: personally, I'd place all exports directly in the `case-assist.indext.ts` file.

File: packages/headless/src/features/case-inputs/case-inputs-actions-loader.ts
Comment (samisayegh) [inline packages/headless/src/features/case-inputs/case-inputs-actions-loader.ts:32]: Suggestion: using a singular `Input` in the names, because the grouping and so the plural is on the actions, not the inputs.

`loadCaseInputsActions` -> `loadCaseInputActions`
`CaseInputsActionCreators` -> `CaseInputActionCreators`

File: packages/headless/src/features/case-fields/case-fields-slice.ts
Comment (samisayegh) [reply packages/headless/src/features/case-fields/case-fields-slice.ts:27]: It might be simpler to manage this UX detail outside of headless, at the UI level.

Generally speaking, we have not embedded UX behaviour at the reducer level. When we have, it's been opinionated with no ability to configure. e.g. clicking a query suggestion updates the search box q.

As long as headless provide the building blocks to a) adjust an input value, and b) trigger a fetch, an implementer can combine them as they wish. This removes the need for
1. Implementers to specify config into headless.
2. Coveo devs to maintain the logic, which can spiral in complexity over time as more options are added to case-assist. Less options delays questions like "what should the behaviour be if option A is `X` and option B is `Y`".


```
<input onChange: autoUpdateClassifications/>

function autoUpdateClassifications(e) {
   inputController.updateValue(e.value);
   inputController.fetchClassifications();
}
```

File: packages/headless/src/case-assist.index.ts
Comment (samisayegh) [inline packages/headless/src/case-assist.index.ts:29]: Related to the other comment, I'd just export the loaders from this file.

File: packages/headless/src/features/case-fields/case-fields-slice.ts
Comment (samisayegh) [inline packages/headless/src/features/case-fields/case-fields-slice.ts:15]: Hmm, I feel we could split this action into `registerCaseField` and `updateCaseField`, and remove the if/else

File: packages/headless/src/features/case-fields/case-fields-actions.ts
Comment (samisayegh) [inline packages/headless/src/features/case-fields/case-fields-actions.ts:32]: Similar to a discussion on another PR, I feel `update` would be more consistent for actions that update state over time.

File: packages/headless/src/features/case-fields/case-fields-slice.ts
Comment (lbergeron) [reply packages/headless/src/features/case-fields/case-fields-slice.ts]: As @nathanlb mentioned, the idea behind the "enabled" flags is to make it easier to the implementer by retrieving the document and field suggestions automatically. We can decide to remove the flag and let the implementer invoke the actions himself, but it adds more complexity on the implementer side. I'm worried about implementers forgetting to update suggestions.

Comment (samisayegh) [reply packages/headless/src/features/case-fields/case-fields-slice.ts]: Going to play devil's advocate, the phrase "worried about implementers forgetting" sounds like there is an opinionated UX we want to push devs towards?

If so, we can do this at the controller level, where e.g. the `updateValue` method dispatches the fetch, similar to the facet controllers. This could make sense if it's the behaviour we expect 99% of implementations to use.

For the 1% that does not want this, the door remains open to override the method and dispatch only the action that updates the input value.

If it's closer to 50-50, then being opinionated would not be the best path.
However, providing the building blocks for devs to compose still might have benefits of minimizing config and code to maintain.

Comment (nathanlb) [reply packages/headless/src/features/case-fields/case-fields-slice.ts]: The planned behaviour is to have the caseInput controller fetch when updateValue is called, but only if the enabled flag is on. The purpose of there being a flag is also that the case form and the case field with case classifications might be on distinct steps of the case submission form. That's how it'll be configured in our cookbook implementation. In that case we don't want classifications to be fetched if there's no use for them yet.

Comment (nathanlb) [reply packages/headless/src/features/case-fields/case-fields-slice.ts]: @samisayegh we'll follow your advice and remove the enable/disable pattern for now. Like I mentioned we'll be putting together a case assist implementation as part of our cookbook and so if the DX reveals some pain points we can re-advise. It would've been perhaps a little more efficient to settle this while reviewing the design document so hopefully we surface these types of questions earlier next time. Thanks for the review! 😁

**Review by lbergeron — APPROVED** [review]

**Pull request merged**
